### PR TITLE
policy: add support for kubernetes endPort

### LINKED ISSUE #16622: Port ranges in network policies
The new `endPort` field for K8s' NetworkPolicies is in [alpha in v1.21](https://kubernetes.io/docs/concepts/services-networking/network-policies/#targeting-a-range-of-ports) and likely to become stable soon. It allows policies to target a range of ports instead of a single port.

Supporting this in Cilium will require changes in the datapath.

Tasks:
* [x] Add port range management mechanism
  * https://github.com/cilium/cilium/pull/20633
* [x] Add support for parsing/embedding port ranges from the API into the policy engine
* [x] Extend datapath for port range support
* [x] Implement policy engine -> datapath map generation logic

### PR DESCRIPTION
**Pull request by rikatz**
- [X] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [X] All code is covered by unit and/or runtime tests where feasible. - Added unit test
- [X] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [X] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [X] Provide a title or release-note blurb suitable for the release notes.
- [ ] Thanks for contributing!

Kubernetes supports endPort field since v1.21, being beta in v1.22 and enabled
by default. This commit adds basic support for parsing endPort field while
still keeps the 40 maximum ports per rule imposed by Cilium parser.

Fixes: #16622
Signed-off-by: Ricardo Pchevuzinske Katz <ricardo.katz@gmail.com>

```release-note
Added support for Kubernetes endPort field / port range
```

### CODE REVIEW

**Comment by nathanjsweet** [comment]
You need a release note, but also you need to [add some integration tests](https://docs.cilium.io/en/v1.10/contributing/testing/) to make this viable.

**Comment by joestringer** [comment]
Oh yeah, I probably should have posted here rather than on the base issue. My latest question was basically whether we think that this implementation is "good enough" for the upstream k8s use cases or not, see more detail here: https://github.com/cilium/cilium/issues/16622#issuecomment-914587235

**Review by aanm — APPROVED** [review]

File: pkg/k8s/network_policy_test.go [hunk 1]
```diff
@@ -700,6 +700,48 @@ func (s *K8sSuite) TestParsePorts(c *C) {
...
+			},
+		},
+	}
+
+	rules, err := ParseNetworkPolicy(netPolicy)
+	c.Assert(err, IsNil)
+	c.Assert(len(rules), Equals, 1)
+	c.Assert(len(rules[0].Egress[0].ToPorts[0].Ports), Equals, 40)
+	c.Assert(len(rules[0].Ingress[0].ToPorts[0].Ports), Equals, 11)
+
```
Comment (christarazi) [inline pkg/k8s/network_policy_test.go:742]: Nit: extra line

**Review by christarazi — APPROVED** [review]

**Comment by rikatz** [comment]
Sorry for the delay folks, day job duties taking me some time.

Will cycle back on this during this week/end :)

**Review by nathanjsweet — APPROVED** [review]

**Review by joestringer — CHANGES_REQUESTED** [review]
Just to make sure we engage with the discussion, I'm setting requested changes here. I think we need to decide whether we think it's useful to have a simpler implementation like this with some limitations or whether we should engage with a more comprehensive solution. [Discussion link](https://github.com/cilium/cilium/issues/16622#issuecomment-914587235).

**Comment by rikatz** [comment]
Hi folks,

I wont have much time to work on this anymore. If someone wants to take over this PR, please feel free, otherwise close it.

Thanks

**Pull request closed without merging**